app/scripts/migrations/140.ts

```diff
@@ -0,0 +1,251 @@
+    log.warn(
+      'Skipping migration: `NetworkController` state not found or is not an object.',
+    );
+    return state;
+  }
+
+  if (
+    !hasProperty(networkControllerState, 'selectedNetworkClientId') ||
+    typeof networkControllerState.selectedNetworkClientId !== 'string'
+  ) {
+    log.warn(
+      'Skipping migration: `NetworkController.selectedNetworkClientId` is not a string.',
+    );
+    return state;
+  }
+
+  if (
+    !hasProperty(networkControllerState, 'networkConfigurationsByChainId') ||
+    !isObject(networkControllerState.networkConfigurationsByChainId)
+  ) {
+    log.warn(
+      'Skipping migration: `NetworkController.networkConfigurationsByChainId` is not an object.',
+    );
+    return state;
+  }
+
+  if (!permissionControllerState || !isObject(permissionControllerState)) {
+    log.warn(
+      'Skipping migration: `PermissionController` state not found or is not an object.',
+    );
+    return state;
+  }
+
+  if (!isObject(permissionControllerState.subjects)) {
+    log.warn(
+      'Skipping migration: `PermissionController.subjects` state is not an object.',
+    );
+    return state;
+  }
+
+  if (
+    !selectedNetworkControllerState ||
+    !isObject(selectedNetworkControllerState)
+  ) {
+    log.warn(
+      'Skipping migration: `SelectedNetworkController` state not found or is not an object.',
+    );
+    return state;
+  }
+
+  if (!isObject(selectedNetworkControllerState.domains)) {
+    log.warn(
+      'Skipping migration: `SelectedNetworkController.domains` state is not an object.',
+    );
+    return state;
+  }
+
+  const currentChainId = getChainId(
+    networkControllerState.selectedNetworkClientId,
+    networkControllerState.networkConfigurationsByChainId,
+  );
+  if (typeof currentChainId !== 'string') {
+    log.warn(
+      'Skipping migration: `NetworkController.selectedNetworkClientId` is not a string.',
+    );
+    return state;
+  }
+
+  const updatedSubjects: string[] = [];
+
+  // Add permission to use the current globally selected network to all Snaps
+  // that have the `endowment:ethereum-provider` permission.
+  const entries = Object.entries(permissionControllerState.subjects) as [
+    string,
+    GenericPermissionControllerSubject,
+  ][];
+
+  permissionControllerState.subjects = entries.reduce<
+    PermissionControllerSubjects<PermissionConstraint>
+  >((accumulator, [key, subject]) => {
+    const permissionKeys = Object.keys(subject.permissions);
+    const needsMigration =
+      permissionKeys.includes(SnapEndowments.EthereumProvider) &&
+      !permissionKeys.includes(PermissionNames.permittedChains);
+
+    if (!needsMigration) {
+      return {
+        ...accumulator,
+        [key]: subject,
+      };
+    }
+
+    updatedSubjects.push(key);
+
+    const caveatValue = addPermittedEthChainId(
+      {
+        requiredScopes: {},
+        optionalScopes: {},
+        isMultichainOrigin: false,
+      },
+      currentChainId as `0x${string}`,
+    );
```

Comment:
I think this logic will have to account for whether or not there are also existing account permissions, because this will overwrite those permissions if so.

Reply:
These account permissions will have been migrated into the CAIP-25 permission already: https://github.com/MetaMask/metamask-extension/pull/27847/files#diff-e63a7581f2f988fe114f58ecfaa796e1f6efc4f8934e5883dfca9204618edb13R402